source/common/grpc/grpc_web_filter.cc

```diff
@@ -0,0 +1,109 @@
+#include "common/grpc/grpc_web_filter.h"
+
+#include <arpa/inet.h>
+
+#include "common/common/base64.h"
+#include "common/http/headers.h"
+
+namespace Envoy {
+namespace Grpc {
+
+const uint8_t GrpcWebFilter::GRPC_WEB_TRAILER = 0b10000000;
+
+GrpcWebFilter::GrpcWebFilter() : is_text_request_(false), is_text_response_(false) {}
+
+GrpcWebFilter::~GrpcWebFilter() {}
+
+// Implements StreamDecoderFilter.
+Http::FilterHeadersStatus GrpcWebFilter::decodeHeaders(Http::HeaderMap& headers, bool) {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type != nullptr &&
+      Http::Headers::get().ContentTypeValues.GrpcWebText == content_type->value().c_str()) {
+    is_text_request_ = true;
+  }
+  headers.insertContentType().value(Http::Headers::get().ContentTypeValues.Grpc);
+
+  const Http::HeaderEntry* accept = headers.get(Http::Headers::get().Accept);
+  if (accept != nullptr &&
+      Http::Headers::get().ContentTypeValues.GrpcWebText == accept->value().c_str()) {
+    is_text_response_ = true;
+  }
+  headers.addStatic(Http::Headers::get().TE, Http::Headers::get().TEValues.Trailers);
+  headers.addStatic(Http::Headers::get().GrpcAcceptEncoding,
+                    Http::Headers::get().GrpcAcceptEncodingValues.Default);
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus GrpcWebFilter::decodeData(Buffer::Instance& data, bool) {
+  if (!is_text_request_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  // Parse application/grpc-web-text format.
+  if (data.length() + decoding_buffer_.length() < 4) {
+    decoding_buffer_.move(data);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t needed =
+      (data.length() + decoding_buffer_.length()) / 4 * 4 - decoding_buffer_.length();
+  decoding_buffer_.move(data, needed);
+  const std::string decoded = Base64::decode(
+      std::string(static_cast<const char*>(decoding_buffer_.linearize(decoding_buffer_.length())),
+                  decoding_buffer_.length()));
+  decoding_buffer_.drain(decoding_buffer_.length());
+  decoding_buffer_.move(data);
+  data.add(decoded);
+  return Http::FilterDataStatus::Continue;
+}
+
+// Implements StreamEncoderFilter.
+Http::FilterHeadersStatus GrpcWebFilter::encodeHeaders(Http::HeaderMap& headers, bool) {
+  if (is_text_response_) {
+    headers.insertContentType().value(Http::Headers::get().ContentTypeValues.GrpcWebText);
+  } else {
+    headers.insertContentType().value(Http::Headers::get().ContentTypeValues.GrpcWeb);
+  }
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus GrpcWebFilter::encodeData(Buffer::Instance& data, bool) {
+  if (!is_text_response_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  // Encodes the response as base64.
```

Comment:
Can you elaborate in some more depth on the transformation that is happening here? My current guess is that we're taking standard gRPC wire format, decoding and deframing, then re-encoding it using the format of gRPC Web? What is the meaning of flags in this context? It's probably not H2 flags, so `decoder_` must have figured out what the delimited message flags are? I think lots of verbose comments would help.